org.librarysimplified.r2.vanilla/src/main/java/org/librarysimplified/r2/vanilla/SR2Controller.kt

```diff
@@ -219,9 +224,19 @@ class SR2Controller private constructor(
     targetIndex: Int,
     atEnd: Boolean
   ) {
+    val previousIndex = this.currentChapterIndex
+
     try {
       val location = this.locationOfSpineItem(targetIndex)
       this.logger.debug("opening location {}", location)
+
+      // Warning: The current chapter must be set before loading the spine location. The
+      // page will send a reading position changed event immediately on load, but before our
+      // callback returns. If we don't update the chapter index first we'll report the wrong
+      // chapter location in our SR2ReadingPositionChanged event. This is fragile and should
+      // be fixed later.
+      this.setCurrentChapter(targetIndex)
```

Comment:
It looks like we'll be expected to save the chapter index and progression value when it comes time to save reading positions. We can probably set both in one command.

Reply:
Yeah, that makes sense to me.